### PR TITLE
[TASK] Update ``doctrine/orm`` to 2.4 for PHP 7 compatibility

### PR DESCRIPTION
**Pull request by aertmann**
Prevents segmentation faults caused by Doctrine.

Excludes support for ``EntityListeners`` annotation introduced for 3.0

### CODE REVIEW

**Comment by aertmann** [comment]
:+1: tests pass locally, travis can't test since it doesn't install the correct doctrine version

File: TYPO3.Flow/Classes/TYPO3/Flow/Persistence/Doctrine/Mapping/Driver/FlowAnnotationDriver.php [hunk 1]
```diff
@@ -11,23 +11,44 @@
...
+use Doctrine\ORM\Mapping\JoinTable;
+use Doctrine\ORM\Mapping\MappingException;
+use Doctrine\ORM\Mapping\NamedQuery;
 use TYPO3\Flow\Annotations as Flow;
+use TYPO3\Flow\Aop\Builder\ClassNameIndex;
+use TYPO3\Flow\Aop\Pointcut\PointcutFilterInterface;
+use TYPO3\Flow\Object\Proxy\Compiler;
+use TYPO3\Flow\Persistence\Doctrine\Mapping\Exception\ClassSchemaNotFoundException;
+use TYPO3\Flow\Reflection\ClassSchema;
+use TYPO3\Flow\Reflection\ReflectionService;
```
Comment (bwaidelich) [inline TYPO3.Flow/Classes/TYPO3/Flow/Persistence/Doctrine/Mapping/Driver/FlowAnnotationDriver.php]: It would be easier to review if we did not mix boy-scouting and migration here

File: TYPO3.Flow/Classes/TYPO3/Flow/Persistence/Doctrine/Mapping/Driver/FlowAnnotationDriver.php [hunk 2]
```diff
@@ -159,19 +189,21 @@ public function loadMetadataForClass($className, \Doctrine\Common\Persistence\Ma
...
         // Evaluate Table annotation
         $primaryTable = array();
         if (isset($classAnnotations['Doctrine\ORM\Mapping\Table'])) {
             $tableAnnotation = $classAnnotations['Doctrine\ORM\Mapping\Table'];
-            $primaryTable['name'] = $tableAnnotation->name;
-            $primaryTable['schema'] = $tableAnnotation->schema;
+            $primaryTable = array(
+                'name' => $tableAnnotation->name,
+                'schema' => $tableAnnotation->schema
+            );
```
Comment (bwaidelich) [inline TYPO3.Flow/Classes/TYPO3/Flow/Persistence/Doctrine/Mapping/Driver/FlowAnnotationDriver.php]: Why this change?

File: TYPO3.Flow/Classes/TYPO3/Flow/Persistence/Doctrine/Mapping/Driver/FlowAnnotationDriver.php [hunk 3]
```diff
@@ -1035,10 +1088,10 @@ public function getRuntimeEvaluationsDefinition()
...
-     * @param \TYPO3\Flow\Aop\Builder\ClassNameIndex $classNameIndex
-     * @return \TYPO3\Flow\Aop\Builder\ClassNameIndex
+     * @param ClassNameIndex $classNameIndex
+     * @return ClassNameIndex
      */
-    public function reduceTargetClassNames(\TYPO3\Flow\Aop\Builder\ClassNameIndex $classNameIndex)
+    public function reduceTargetClassNames(ClassNameIndex $classNameIndex)
     {
         return $classNameIndex;
     }
```
Comment (bwaidelich) [inline TYPO3.Flow/Classes/TYPO3/Flow/Persistence/Doctrine/Mapping/Driver/FlowAnnotationDriver.php]: If I did not miss anything the only functional changes are those of lines 534/578, 596/636 and everything after line 820, right?

File: TYPO3.Flow/Classes/TYPO3/Flow/Persistence/Doctrine/Mapping/Driver/FlowAnnotationDriver.php (hunk 1 above)
Comment (aertmann) [inline TYPO3.Flow/Classes/TYPO3/Flow/Persistence/Doctrine/Mapping/Driver/FlowAnnotationDriver.php]: totally agree, made it much harder to backport. and heh right back at you, it's originally your change ;) https://github.com/neos/flow-development-collection/commit/5483b67eef8a161049f5690300da94e2d3a16b18

File: TYPO3.Flow/Classes/TYPO3/Flow/Persistence/Doctrine/Mapping/Driver/FlowAnnotationDriver.php (hunk 2 above)
Comment (aertmann) [inline TYPO3.Flow/Classes/TYPO3/Flow/Persistence/Doctrine/Mapping/Driver/FlowAnnotationDriver.php]: part of ported commit..

File: TYPO3.Flow/Classes/TYPO3/Flow/Persistence/Doctrine/Mapping/Driver/FlowAnnotationDriver.php (hunk 1 above)
Comment (bwaidelich) [inline TYPO3.Flow/Classes/TYPO3/Flow/Persistence/Doctrine/Mapping/Driver/FlowAnnotationDriver.php]: Well, Karstens Patch-Set https://review.typo3.org/#/c/36873/2  ;)

File: TYPO3.Flow/Classes/TYPO3/Flow/Persistence/Doctrine/Mapping/Driver/FlowAnnotationDriver.php (hunk 2 above)
Comment (bwaidelich) [inline TYPO3.Flow/Classes/TYPO3/Flow/Persistence/Doctrine/Mapping/Driver/FlowAnnotationDriver.php]: but it's not in 5483b67 !?

Comment (bwaidelich) [inline TYPO3.Flow/Classes/TYPO3/Flow/Persistence/Doctrine/Mapping/Driver/FlowAnnotationDriver.php]: ah of course it is, nevermind

**Comment by kdambekalns** [comment]
:-1: The functional tests pass locally, indeed. But the Neos functional tests fail even with PHP 5.6. So if this is merged, Neos 1.2 will be broken, it seems.

**Comment by aertmann** [comment]
@kdambekalns: yeah the Neos tests need to be adjusted, figured that was done for 2.0 as well

**Comment by kdambekalns** [comment]
It is done for Neos 2.0,  but Flow 2.3 is for Neos 1.2…

**Comment by aertmann** [comment]
@kdambekalns: I think you miss understood, tried to say that there is something that needs to be adjusted for the 1.2 Neos branch to make it compatible with ORM 2.4. But since that is already done for for Neos 2.0 that shouldn't be a problem. Planned to do that after this is merged.

**Comment by kdambekalns** [comment]
Ok. Please do, because this should be merged together with the other fix, or it'll break tests in between… and it might even break setups for the less cautious out there.

**Comment by aertmann** [comment]
broken functional tests fixed with #251

**Comment by aertmann** [comment]
merging this since there are no objections beside the one from Karsten which has been fixed

**Comment by aertmann** [comment]
trivial rebase

**Comment by kitsunet** [comment]
nice!

**Comment by kdambekalns** [comment]
…and broken. Or is that a temporary glitch caused by some outdated dependencies?

**Comment by kitsunet** [comment]
Didn't @aertmann say there is some test adaption needed afterwards?

**Comment by aertmann** [comment]
don't know why the first test failed, but they all pass now without doing anything.. seems like it didn't install the right doctrine dependency

**Comment by kitsunet** [comment]
yes, yes, that happened to me on 3.0 as well... all good

**Comment by kitsunet** [comment]
See also https://github.com/neos/neos-development-collection/commit/077c3116548619a7b63be7781ca283a681898383

**Comment by kdambekalns** [comment]
Nice. All good then. :8ball:
